### PR TITLE
add python 3.11 support

### PR DESCRIPTION
**Pull request by tvdboom**
# Describe the changes you've made

Add support for python 3.11.

Additionnaly, fix a bug in `test_classification_custom_metric`.

# Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by tvdboom** [comment]
@Yard1 @ngupta23 Can you guys have a look? It's unclear to me which dependency is making it fail here. Also, why aren't we supporting `pandas>2.0`. I think we should also try to support that for next release.

**Comment by moezali1** [comment]
numba <0.58 is not supported on python 3.11.

image

**Comment by amotl** [comment]
Hi there,

first things first: Thanks a stack for conceiving and maintaining PyCaret, and for working on this patch.

We have been tripped by missing support for Python 3.11 when trying to bring in support for MLflow.
- https://github.com/crate-workbench/mlflow-cratedb/pull/45#pullrequestreview-1703789317

With kind regards,
Andreas.

**Comment by tvdboom** [comment]
Hi @amotl, we plan to support 3.11 with next release but we have some issues with dependencies. All help is welcome.

**Comment by amotl** [comment]
Dear Mavs,

GH-3813 might support this patch a bit. Apparently, more work needs to be done to fully support Python 3.11 on all platorms, as subsequent PRs will demonstrate. But at least, it's a start, and may already satisfy a good share of users, when combined with this patch, and published with an upcoming release.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Hi again,

we cleaned up our act, and submitted two followup patches to conclude Python 3.11 support.

- GH-3814
- GH-3815

With kind regards,
Andreas.

**Comment by amotl** [comment]
Thank you for integrating our patches. 💯

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -88,138 +97,17 @@ jobs:
...
-  #         python --version
-  #         which python
-  #         pip list
-  #     - name: Remove tests
-  #       run: |
-  #         remove-item tests/* -Include @('test_clustering.py', 'test_classification_tuning.py','test_classification_plots.py','test_regression_plots.py', 'test_regression_tuning.py', 'test_time_series_tune_grid.py', 'test_time_series_tune_random.py', 'test_create_api.py', 'test_create_docker.py', 'test_drift_report.py', 'test_eda.py', 'test_time_series_plots.py', 'test_time_series_utils_plots.py')
-  #         remove-item tests/benchmarks/*
-  #     - name: Test with pytest
-  #       run: pytest --durations=0
+        run: pytest --durations=0 -m "not (benchmark or plotting or tuning_random)"
```
[the remark this answers is not in the document]
Comment (ngupta23) [reply .github/workflows/test.yml:101]: OK, I see that you have separated out random and grid. So maybe add grid in addition to random here also.

File: tests/test_time_series_plots.py [hunk 2]
```diff
@@ -19,6 +20,9 @@
 pytestmark = pytest.mark.filterwarnings("ignore::UserWarning")
 os.environ["PYCARET_TESTING"] = "1"
 
+if sys.platform == "win32":
```
Comment (ngupta23) [inline tests/test_time_series_plots.py:23]: Why are we skipping these on Windows?

File: .github/workflows/test.yml (hunk 1 above)
Comment (Yard1) [inline .github/workflows/test.yml:101]:
```suggestion
        run: pytest --durations=0 -m "not (benchmark or plotting or tuning_random or tuning_grid)"
```

**Review by Yard1 — APPROVED** [review]

File: tests/test_time_series_plots.py (hunk 2 above)
Comment (amotl) [reply tests/test_time_series_plots.py:23]: I think those are the corresponding references.

- https://github.com/pycaret/pycaret/pull/3814#issuecomment-1794003113
- https://github.com/ray-project/tune-sklearn/issues/283

**Pull request merged**